### PR TITLE
Simple clouds

### PR DESCRIPTION
**Pull request by penmaher**
This simple cloud scheme was implemented for using Isca in the second phase of the SPOOKIE model intercomparison. It then went on to form the base for Qun Liu's paper: https://gmd.copernicus.org/articles/14/2801/2021/

@lqxyz: requesting your review for your expertise on this topic and to ensure I don't put anything on the master that will make your merge more difficult.

@RuthG requesting your input (don't need a full review unless you want to) as the test cases for the cloud scheme rely on your new era topography file (which has improved the Gibbs ripples). This file will is not part of this PR.

@sit23 requesting your input (don't need a full review unless you want to) as I change the socrates test cases.

### CODE REVIEW

**Comment by sit23** [comment]
Hi @penmaher - thanks for this. I'll take a look at the test case changes, as you requested.

**Comment by penmaher** [comment]
Following Ruth's suggestion i have added the .nc file used in the test cases

File: src/atmos_param/cloud_simple/cloud_simple.F90
Comment (lqxyz) [inline src/atmos_param/cloud_simple/cloud_simple.F90:75]: Hi Penny, it looks like you haven't defined `nml_unit` yet, and perhaps define it as an integer variable in line `68`.

File: src/atmos_param/cloud_simple/cloud_simple.F90
Comment (lqxyz) [inline src/atmos_param/cloud_simple/cloud_simple.F90:81]: Hi Penny, could you add some error checks for the namelist so as to find the misspelled namelist? For example:
```
     integer :: io, ierr, nml_unit, stdlog_unit

#ifdef INTERNAL_FILE_NML
    read(input_nml_file, nml=cloud_simple_nml, iostat=io)
    ierr = check_nml_error(io, 'cloud_simple_nml')
#else
    if (file_exist('input.nml')) then
      nml_unit = open_namelist_file()
      ierr = 1
      do while (ierr /= 0)
        read(nml_unit, nml=cloud_simple_nml, iostat=io, end=10)
        ierr = check_nml_error(io, 'cloud_simple_nml')
      enddo
10    call close_file(nml_unit)
    endif
#endif
    stdlog_unit = stdlog()
    write(stdlog_unit, cloud_simple_nml)
```

File: src/atmos_param/cloud_simple/cloud_simple.F90
Comment (lqxyz) [inline src/atmos_param/cloud_simple/cloud_simple.F90:206]: Delete one blank line?

**Review by lqxyz — COMMENTED** [review]
Hi @penmaher, thank you very much for coding the SPOOKIE cloud scheme. I have reviewed the file `cloud_simple.F90` and all the logics look good to me. I just made some comments on the namelist check, some tidy-ups and the possible simplification of the nested loops (this needs more work, so just neglect it if it's not easy). I will review the other files later.

File: src/atmos_param/socrates/interface/socrates_interface.F90
Comment (lqxyz) [inline src/atmos_param/socrates/interface/socrates_interface.F90:64]: Hi @penmaher, thank you very much for adding more outputs about the clear-sky fluxes, which makes the calculation of cloud radiative effect much easier.

**Review by lqxyz — COMMENTED** [review]
Hi @penmaher, many thanks for adding the clear-sky fluxes in `Socrates` interface, which makes the calculation of cloud radiative effect much easier! Also, the simple cloud and Socrates interfaces in `idealized_moist_phys.F90` look good to me and the file has become cleaner after your clean-up to the old comments.

**Review by penmaher — COMMENTED** [review]
Hey @Qun. Thanks for the excellent review. I have made these changes but I am now doing local testing to make sure the model still compiles etc. I will push the changes when I am confident I have not broken anything.

**Comment by penmaher** [comment]
@lqxyz the runscripts needed some updating and the code changes now compile and run too. Should be good for you to review my changes.

File: src/atmos_param/cloud_simple/cloud_simple.F90
Comment (lqxyz) [inline src/atmos_param/cloud_simple/cloud_simple.F90:306]: Delete the extra space after `MIN(`?

File: src/atmos_spectral/driver/solo/idealized_moist_phys.F90
Comment (lqxyz) [inline src/atmos_spectral/driver/solo/idealized_moist_phys.F90:212]: Thank you very much for making the code aligned and much nicer to read.

File: src/atmos_spectral/driver/solo/idealized_moist_phys.F90
Comment (lqxyz) [inline src/atmos_spectral/driver/solo/idealized_moist_phys.F90:589]: More indentation for this line?

File: src/atmos_spectral/driver/solo/idealized_moist_phys.F90
Comment (lqxyz) [inline src/atmos_spectral/driver/solo/idealized_moist_phys.F90:776]: Perhaps more indentation for this line?

File: src/atmos_spectral/driver/solo/idealized_moist_phys.F90
Comment (lqxyz) [inline src/atmos_spectral/driver/solo/idealized_moist_phys.F90:1058]: This looks great to me.

File: src/atmos_param/cloud_simple/cloud_simple.F90
Comment (lqxyz) [inline src/atmos_param/cloud_simple/cloud_simple.F90:216]: Perhaps add more spaces around the minus sign? Changing the `(KELVIN-temp)` to `(KELVIN - temp)`?

**Review by lqxyz — COMMENTED** [review]
Hi @penmaher, thank you very much for the updates to the codes, which makes the code cleaner and easier to read. All the changes look good to me and only several minor changes are still needed.

**Comment by penmaher** [comment]
Thanks for the second review Qun. All suggested changes have been made and pushed

**Comment by lqxyz** [comment]
Hi @penmaher, the changes look good to me, but do you know why the CI test [(job #120.2)](https://travis-ci.com/github/ExeClim/Isca/jobs/517587548#L936) failed? It seems that the `idealized_moist_phys.F90` could not find the `cloud_simple_mod`, but the `cloud_simple.F90` has been specified in [`path_names`](https://github.com/penmaher/Isca/blob/simple_clouds/src/extra/model/socrates/path_names#L14). But it's a little weird as this branch could be compiled smoothly when I test it locally.

**Comment by dennissergeev** [comment]
@lqxyz could you try checking out this branch in a completely new directory and compiling it from scratch locally?

One error I noticed is in the [compile_mppn.sh](https://github.com/penmaher/Isca/blob/simple_clouds/postprocessing/compile_mppn.sh#L13) script. Line 13 should be changed to 
```bash
source $GFDL_ENV
```

**Comment by penmaher** [comment]
Thanks for the suggestion @dennissergeev. I downloaded a new repo of this branch, it compiles but does not run. I am investigating. This is weird as it runs fine in the clone I did the dev on (though I have not cleaned the compile directory).   

In terms of the suggested change to  compile_mppn.sh script. Line 13, I think the full path and filename are needed, not just the filename, or at least it is in the current way I have the environment variables set.

**Comment by dennissergeev** [comment]
> In terms of the suggested change to compile_mppn.sh script. Line 13, I think the full path and filename are needed, not just the filename, or at least it is in the current way I have the environment variables set.

For the automatic tests, `GFDL_ENV` already includes the full path, see https://github.com/penmaher/Isca/blob/simple_clouds/.travis.yml#L12

It could be changed if you think it's not appropriate, but for now just try changing your `GFDL_ENV` to include the path.

**Comment by sit23** [comment]
Hi @penmaher - thanks for this. The socrates test cases look fine to me, and it's nice to have all the options included. 

Did you want to include one of the clouds on cases in the trip tests? It might be a good idea.

I was also wondering about the RRTM modifications you've included here. Did we actually get those to work in the end? And if not, is it worth excluding them from this? 

As a final note, I'm guessing that you'll run the trip tests on this PR once all the changes are in? No point in running them sooner, perhaps.

**Comment by sit23** [comment]
And to respond to the point about the `compile_mppn.sh`, I think the way most of us have the $GFDL_ENV set up means that the full path is required. But you're right @dennissergeev that it might be better to change this. It would be quite a big change though. We could probably compromise where the front-end would check whether the full path or just the GFDL_ENV file exists, and then uses this one?

**Comment by dennissergeev** [comment]
@sit23 OK, I see. To be honest I don't remember why I set it up like that in `.travis.yml`. 
@penmaher or @lqxyz could you change the file accordingly please (lines 12 and 40)?

**Comment by lqxyz** [comment]
Hi @dennissergeev, I have tried to clone this branch to a new directory, and it could compile correctly. And it still works after modifying the `GFDL_ENV` to its full path in `.bashrc`, and changing `source $GFDL_BASE/src/extra/env/$GFDL_ENV` to `source $GFDL_ENV` in `compile_mppn.sh`.

**Comment by dennissergeev** [comment]
@lqxyz thanks for testing this.

Even though this error seems to be unaffected by `$GFDL_ENV`, could you please add a commit with the changes to `.travis.yml` as I described in my previous comment. Then we'll check the travis test log again.

**Comment by dennissergeev** [comment]
So the compilation failed again, but at least the bug with `$GFDL_ENV` is now fixed.
The reason for the compilation error seems to be that `cloud_simple` is not being compiled prior to `idealized_moist_phys`...

**Comment by dennissergeev** [comment]
@lqxyz @penmaher check that you added `cloud_simple` to _all_ path_names:
```bash
# in isca's top-level directory
find . -name path_names -type f -print0 | xargs -0 grep cloud_simple
```

**Comment by sit23** [comment]
I've run the trip tests on the current head of the master (8a715ca) vs the last 'clean' commit on this simple clouds branch (8c39828, i.e. the last commit before Penny started hunting for bugs). The results all pass:
```
Results for all of the test cases ran comparing 8c39828 and 8a715ca are as follows...
axisymmetric : pass
bucket_model : pass
frierson : pass
giant_planet : pass
held_suarez : pass
MiMA : pass
realistic_continents_fixed_sst : pass
realistic_continents_variable_qflux : pass
socrates_aquaplanet : pass
top_down_test : pass
variable_co2_grey : pass
variable_co2_rrtm : pass
ape_aquaplanet : pass
Congratulations, all tests have passed
```
This is clearly good news, as it means the simple_clouds branch can be merged in cleanly. However, @penmaher did almost exactly the same tests and found failures, so there's still a bit of mystery here. But, on the face of it, the tests are passing. I think if we can revert the commits @penmaher did when looking for bugs then we can happily merge this in.

**Comment by sit23** [comment]
OK - I've reverted all of @penmaher's bug-finding commits, and there are no differences between the current HEAD (e70400c) and her last clean commit 8c39828 (i.e. `git diff e70400c 8c39828` returns nothing). I'll run the trip tests on the current HEAD just to be sure.

**Comment by sit23** [comment]
In re-examining the trip-tests, I think I have identified the reason for @penmaher's trip test failures when I get the trip test as clean. In this p/r the namelist parameters for the `socrates_aquaplanet.py` experiment are changed.  This means that the results from this experiment will change depending on whether it is run with the new or old configuration. Now the trip tests are designed to run each experiment with a set of commits with a consistent namelist. However, the situation can arise where someone runs the trip test with a given commit using a given version of the test cases, but then the test cases are changed, and then the trip tests are run again using one of the same commits as before. In that case, the data from the previous run (using the old namelist) is not overwritten, but the new commit uses the new namelist to run. So the trip tests will appear to fail, but only because they have been run with different namelists. I have therefore added a check to the trip tests to make sure that two experiments that are being compared are run with the same namelist values.

I managed to reproduce one of @penmaher's failed trip tests, and this was the reason for it in my case, so I'm guessing it was the case for @penmaher too. On that basis, I have updated the trip tests in this p/r to make sure this doesn't happen again. I also now believe the trip tests to be clean for this p/r. I'm now running the trip tests for the head of this branch vs the current master. Once those are done and clean, I'm happy to merge this in.

**Comment by sit23** [comment]
OK, here are the results of the trip tests for the current head of the master () and the current head of the simple clouds branch ():
```
Results for all of the test cases ran comparing eaa2dc3 and 8a715ca are as follows...
axisymmetric : pass
bucket_model : pass
frierson : pass
giant_planet : pass
held_suarez : pass
MiMA : pass
realistic_continents_fixed_sst : pass
realistic_continents_variable_qflux : pass
socrates_aquaplanet : pass
socrates_aquaplanet_cloud : fail
top_down_test : pass
variable_co2_grey : pass
variable_co2_rrtm : pass
ape_aquaplanet : pass
Nightmare, some tests have failed
```
All of the tests pass apart from the `socrates_aquaplanet_cloud` case. But this is because it requires the use of the cloud module, and this is not present in the commit at the current head of the master (), and so that model run fails. I am therefore happy that this code passes the necessary tests, and can be merged in. I will complete my review to allow this to happen.

File: exp/test_cases/socrates_test/socrates_aquaplanet.py
Comment (sit23) [inline exp/test_cases/socrates_test/socrates_aquaplanet.py:94]: It is noted that this change will change the result of the test case. But this is fine as it will certainly be an obvious change.

Comment (sit23) [reply exp/test_cases/socrates_test/socrates_aquaplanet.py:94]: No action is required on that basis.

**Review by sit23 — APPROVED** [review]
All fine as far as I can see. Trip tests are passing. Happy for this to be merged in.

**Comment by dennissergeev** [comment]
Nice! I guess the next step is to add documentation for this scheme (see #147)

**Comment by sit23** [comment]
Agreed @dennissergeev - perhaps this can wait until after @lqxyz has merged his SimCloud code in?

**Pull request merged**
